### PR TITLE
Feat/email sender service

### PR DESCRIPTION
**Pull request by satello**
- Template for email sending microservices

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -13,7 +13,7 @@
     "commitmsg": "kleros-scripts commitmsg",
     "cz": "kleros-scripts cz",
     "invoke": "env-cmd ./.env serverless invoke -l -f",
-    "build": "docker run --rm -v $PWD:/data -w /data node:8 npm rebuild scrypt",
+    "build": "docker run --rm -v $PWD:/data -w /data lambci/lambda:build-nodejs8.10 npm rebuild scrypt",
```
Comment (epiqueras) [inline package.json:16]: Why the change?

File: serverless.yml [hunk 2]
```diff
@@ -31,6 +31,10 @@ custom:
           path: doges-on-trial,
           description: 'Microservices for the Doges on Trial dapp.',
         }
+      - {
+          path: email-sender,
```
Comment (epiqueras) [inline serverless.yml]: Makes more sense to call this:

event-handlers
Microservices that handle event logs from the blockchain.

File: serverless.yml [hunk 3]
```diff
@@ -62,18 +66,18 @@ custom:
           contentType: application/json,
           schema: '${file(src/global/models/user-settings.json):patch.response}',
         }
-      - {
-          name: PostDogesOnTrialDogeImagesRequest,
-          description: 'The request for postDogesOnTrialDogeImages.',
```
Comment (epiqueras) [inline serverless.yml]: Why is this deleted?

File: serverless.yml [hunk 4]
```diff
@@ -62,18 +66,18 @@ custom:
...
-          schema: '${file(src/doges-on-trial/models/doge-images.json):post.request}',
-        }
       - {
           name: PostDogesOnTrialDogeImagesResponse,
           description: 'The response for postDogesOnTrialDogeImages.',
           contentType: application/json,
           schema: '${file(src/doges-on-trial/models/doge-images.json):post.response}',
         }
+      - {
+          name: PostEmailSenderRequest,
```
Comment (epiqueras) [inline serverless.yml]: This should be generic for event-handlers handlers. You might do something else besides sending an email.

"PostEventHandlersHandlerRequest".

File: serverless.yml [hunk 5]
```diff
@@ -173,3 +177,19 @@ functions:
             responseModels: { application/json: ErrorResponse },
           },
         ]
+  postEmailSender:
```
Comment (epiqueras) [inline serverless.yml]: "postEventHandlersExampleEmail"

Real handlers can be: "postEventHandlersKlerosDrawEmail".

File: serverless.yml [hunk 6]
```diff
@@ -173,3 +177,19 @@ functions:
...
+    handler: src/email-sender/email-sender.post
```
Comment (epiqueras) [inline serverless.yml]: src/event-handlers/example-email.post

File: src/email-sender/email-sender.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
+/**
+ * Example email sending lambda function that works with events.kleros.io.
+ * Replace the email lookup with your subscribers for an event and replace your sendgrid template.
+ */
+const sgMail = require('@sendgrid/mail')
+const Web3 = require('web3')
```
Comment (epiqueras) [inline src/email-sender/email-sender.js]: See utils/web3 and other files for how this is done.

File: src/email-sender/email-sender.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    return callback(null, {
+      statusCode: 400,
+      headers: { 'Access-Control-Allow-Origin': '*' },
+      body: JSON.stringify({
+        error: 'No event log passed in body.'
+      })
+    })
+  }
+
+  // FIXME determine which email address you want to use. This example uses the caller of tx that kicks off the event
```
Comment (epiqueras) [inline src/email-sender/email-sender.js]: Don't use FIXME if it's not something that's broken. Use `TODO:` or something like that.

File: src/email-sender/email-sender.js [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+    Key: { address: { S: transaction.from } },
+    TableName: 'user-settings',
+    AttributesToGet: ['email']
+  })
+
+  let emailAddress
+  if (item && item.Item && item.Item.email) emailAddress = item.Item.email.S
+
+  if (emailAddress == null)
+    return callback(null, {
```
Comment (epiqueras) [inline src/email-sender/email-sender.js]: The fact that there is no email should not be reported as an error to the server calling the hook and the status code should be a 204.

File: src/email-sender/email-sender.js [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+      statusCode: 400,
+      headers: { 'Access-Control-Allow-Origin': '*' },
+      body: JSON.stringify({
+        error: 'No registered email address.'
+      })
+    })
+
+  // Sendgrid
+  sgMail.setApiKey(process.env.SENDGRID_API_KEY)
```
Comment (epiqueras) [inline src/email-sender/email-sender.js]: I suggest making a util for this like with web3 and dynamoDB. It will be used by all handlers.

File: src/email-sender/email-sender.js [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+      name: 'Kleros',
+      email: 'contact@kleros.io'
+    },
+    templateId: process.env.SENDGRID_TEMPLATE_ID,
+    dynamic_template_data: {
+      subject: 'Test Email Update',
+      eventName: body.eventName
+    }
+  }
+  sgMail.send(msg)
```
Comment (epiqueras) [inline src/email-sender/email-sender.js]: Is this not async?

File: src/email-sender/email-sender.js [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+  sgMail.setSubstitutionWrappers('{{', '}}')
+
+  // FIXME Add your email template and message here
+  const msg = {
+    to: emailAddress,
+    from: {
+      name: 'Kleros',
+      email: 'contact@kleros.io'
+    },
+    templateId: process.env.SENDGRID_TEMPLATE_ID,
```
Comment (epiqueras) [inline src/email-sender/email-sender.js]: This does not need to be an env var. There's not a different one for staging/production and it is not encrypted.

File: src/email-sender/email-sender.js [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    dynamic_template_data: {
+      subject: 'Test Email Update',
+      eventName: body.eventName
+    }
+  }
+  sgMail.send(msg)
+
+  callback(null, {
+    statusCode: 201,
```
Comment (epiqueras) [inline src/email-sender/email-sender.js]: 201 is resource created. Should be 204.

File: src/email-sender/models/email-sender.json [hunk 14]
```diff
@@ -0,0 +1,22 @@
+{
+  "post": {
+    "request": {
+      "type": "object",
+      "properties": {
+        "payload": {
+          "type": "object",
+          "properties": {
+            "eventName": {
+              "type": "string",
```
Comment (epiqueras) [inline src/email-sender/models/email-sender.json]: This schema does not match what the code expects.

File: README.md [hunk 15]
```diff
@@ -14,9 +14,14 @@
 
 Collection of Kleros microservices.
 
-## Other Scripts
+## Deploying a new microservice
 
-- `yarn run prettify` - Apply prettier to the entire project.
-- `yarn run lint` - Lint the entire project's .js files.
```
Comment (epiqueras) [inline README.md]: Why is this deleted?

File: README.md [hunk 16]
```diff
@@ -14,9 +14,14 @@
...
-- `yarn run lint --fix` - Fix fixable linting errors in .js files.
-- `yarn run cz` - Run commitizen.
+1) Add your service to `serverless.yml`.
+
+2) Get AWS credentials from dashboard and set environment variables `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`.
+
+3) `yarn build` to rebuild problematic dependencies with docker.
+
+4) `yarn deploy:staging` will deploy your new staging microservice for you to test with. `yarn deploy:function <NEW_SERVICE>` can be used after it has been deployed a first time to re-deploy just your function.
```
Comment (epiqueras) [inline README.md]: We should not try to summarize serverless docs in here. We should show a list of our custom scripts, what they do, and link to serverless's website when appropriate.

**Review by epiqueras — CHANGES_REQUESTED** [review]

File: README.md (hunk 15 above)
Comment (satello) [reply README.md]: Doesn't seem useful. They are just the kathari scripts which are the same in all of the projects so not sure why they would be a part of this README

File: README.md (hunk 16 above)
Comment (satello) [reply README.md]: I suppose. My thinking was it makes things easier on our devs if they are completely unfamiliar with the repo so they know where to start at the very least. It's not really going too in depth either, mostly just pointing out that those scripts exist

File: package.json (hunk 1 above)
Comment (satello) [reply package.json:16]: I still had C++ binding errors with just node:8. This is an environment that is supposed to mimic lambda as closely as possible

File: serverless.yml (hunk 3 above)
Comment (satello) [reply serverless.yml]: I don't know... shouldn't be. Maybe I cut instead of copied. I'll put it back

File: serverless.yml (hunk 4 above)
Comment (satello) [reply serverless.yml]: My thinking is this is basically an example service. I played with making some generic and extensible but I think it just makes things more complicated if anyone has event handling that is mildly out of the ordinary. So I mostly set this up so that people can quickly copy/paste and edit to fit their own use case

File: src/email-sender/email-sender.js (hunk 7 above)
Comment (satello) [reply src/email-sender/email-sender.js]: I don't see why that is necessary. What if a project wants to use a different network on staging? It seems limiting to me. And I couldn't tell for sure what the network it was since its encrypted in the secrets file, which also seems unnecessary

File: src/email-sender/email-sender.js (hunk 8 above)
Comment (satello) [reply src/email-sender/email-sender.js]: ok

File: src/email-sender/email-sender.js (hunk 9 above)
Comment (satello) [reply src/email-sender/email-sender.js]: The function essentially is unable to do it's task if there is no email address so I think an error is appropriate. If it is a 204 the server would have no way to know if the email had been sent. The server can choose to ignore the error (the event service does)

File: serverless.yml (hunk 2 above)
Comment (satello) [reply serverless.yml]: That seems too simplistic of a name. It should incorporate email somewhere in the name, as that is the primary purpose of the handler. What about `event-email-handler`

File: src/email-sender/email-sender.js (hunk 7 above)
Comment (epiqueras) [reply src/email-sender/email-sender.js]: We always use Kovan for staging because it's the fastest, and you currently have no way of providing the provider URL, which should also be encrypted.

File: src/email-sender/email-sender.js (hunk 9 above)
Comment (epiqueras) [reply src/email-sender/email-sender.js]: Not having an email set is not an error, it is a valid state in which the user has not set one yet or has opted to not receive notifications.

When the servers sends the event, it has no way of knowing who should receive the email and wether that party has an email set or not. All it knows is that it has to pass on the event. Not sending the email is a valid outcome.

File: README.md (hunk 15 above)
Comment (epiqueras) [reply README.md]: They are not exactly the same in all projects, and it is a good practice to document the package json scripts.

File: README.md (hunk 16 above)
Comment (epiqueras) [reply README.md]: That's what the section for documenting all the scripts is for.

File: src/email-sender/email-sender.js (hunk 9 above)
Comment (satello) [reply src/email-sender/email-sender.js]: I think that really depends on what project the email is being sent for. Since this one isn't tied to a specific project the way I see it, it is more general purpose. And it's better to return an error than silently fail. Returning an error status code shouldn't be a problem for a server that doesn't expect all emails to be able to be sent and is useful for those that do expect it

File: src/email-sender/email-sender.js (hunk 7 above)
Comment (satello) [reply src/email-sender/email-sender.js]: Why does it need to be encrypted? I was just using an env var set in the lambda dashboard and it's just using https://kovan.infura.io?

Comment (satello) [reply src/email-sender/email-sender.js]: Also, is it even possible for me to add things to KMS? Looks like you have the key that is setup with microservices right now. Even if it's more manual to set the env vars via the dashboard it's pretty trivial for everyone to do that for their own services vs dealing with key management to create them as encrypted constants. Also brings up the issue if its even good to have, for example, one encrypted set of API keys for all sendgrid calls. It seems safer to have a separate key for each project so if one starts getting attacked it can be revoked without bringing down every other service that depends on them

File: README.md (hunk 16 above)
Comment (satello) [reply README.md]: Ok i'll add them there. I still think it's nice to have a quick walkthrough. I am sure everyone can figure it out on their own but its a pretty minimal thing that can save them some time at least if they are completely unfamiliar. It doesn't have to live here I guess. I can also just post a quick how to on slack or something

Comment (epiqueras) [reply README.md]: Stepped documentation is good when it is thorough and clear.
Stuff like

>Get AWS credentials from dashboard and set environment variables `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`.

Is very ambiguous and can lead to more questions than answers.

File: src/email-sender/email-sender.js (hunk 7 above)
Comment (epiqueras) [reply src/email-sender/email-sender.js]: That Kovan endpoint will soon be deprecated or severely limited. Infura is becoming a paid service that we will have to pay quotas for. Lambda dashboard env vars are not secure according to their docs and they are meant to be used for configuration variables not API keys.

Yes you can use KMS. https://www.npmjs.com/package/serverless-kms-secrets

We can still encrypt multiple keys if that is desired.

File: src/email-sender/email-sender.js (hunk 9 above)
Comment (epiqueras) [reply src/email-sender/email-sender.js]: That's exactly why it's not an error, because it's generic. The server is just calling an event handler, it does not expect an email to be sent or not. All it expects is for the event to be handled appropriately. Not sending the email can be proper handling. `400` is for malformed requests. A request with an event that does not have an email related to it, is not malformed.

File: src/email-sender/email-sender.js (hunk 7 above)
Comment (satello) [reply src/email-sender/email-sender.js]: AWS encrypts env vars by default. I don't see anything about it being insecure to store API keys there but maybe you are right

Comment (epiqueras) [reply src/email-sender/email-sender.js]: That only happens when you set them specifically for your function in the lambda UI. We would have to do it manually for every service in the repo, very fragile, that's why I created the current setup.

File: src/email-sender/email-sender.js (hunk 9 above)
Comment (satello) [reply src/email-sender/email-sender.js]: I still don't agree that it's not an error but I compromised. Returned 204 but added a body that gives the boolean `sent` and a `reason` so a concerned party still has a way to determine if the request ended up sending an email or not

Comment (epiqueras) [reply src/email-sender/email-sender.js]: Yeah, that's a good approach. But, the code should be 200 then, because you have a body now.
